default.nix

```diff
@@ -78,29 +83,50 @@ stdenv.mkDerivation rec {
            !elem (baseNameOf path) [".git" "result" "bin" "_build" "_build_ci"]) ./.;
 
   preConfigure = ''
+    patchShebangs tools/
     patchShebangs dev/tools/
+    export COQ_CONFIGURE_PREFIX=$out
+  '' + optionalString doInstallCheck ''
+    patchShebangs test-suite/
+  '' + optionalString buildDoc ''
+    patchShebangs doc/tools/
   '';
 
-  prefixKey = "-prefix ";
+  createFindlibDestdir = !shell;
 
-  buildFlags = [ "world" "byte" ] ++ optional buildDoc "doc-html";
+  configurePhase = ''
+    runHook preConfigure
+    runHook postConfigure
+  '';
 
-  installTargets =
-    [ "install" "install-byte" ] ++ optional buildDoc "install-doc-html";
+  makefile = "Makefile.dune";
 
-  createFindlibDestdir = !shell;
+  buildFlags = [ "voboot" ];
 
-  postInstall = "ln -s $out/lib/coq $OCAMLFIND_DESTDIR/coq";
+  postBuild = ''
+    for package in ${toString packages}
+    do
+      dune build $package.install
```

Comment:
If you are building for the final user you should instead use `dune build -p package` which enables the "release" profile. I am not sure how Nix handles Dune packages [coq will soon be 100% standard and don't require `voboot`], but you may want to have a flag for the build to happen in release or in dev mode

See https://dune.readthedocs.io/en/latest/usage.html?highlight=profile#invocation-from-opam for more details. In particular, `-p` requires that the depending packages are system-installed, so indeed that's a tricky tradeoff.

Reply:
This means that the nix build should be split: one per package. Otherwise, we cannot build say `coqide-server` until `coq` is installed.

Reply:
That's what OPAM does, if `coqide` depends on `coqide-server` it will build an install that first.

Note that Nix doesn't have to do that, the composed build is still sound, however you have to be careful to install all the stuff you've build [-p gives you this sanity check as it ignores the rest of packages in the tree].

What would be the best Nix model? I'm sure it would be nice to add native Nix workflow to dune.

Reply:
I would be very much in favor of splitting Coq in as many Nix packages as there are opam ones. Ideally, we would even split the source tree in a way that avoids rebuilding the main Coq package if only the doc changes.

Reply:
Note that I think that for the CI build, we should build in the way this PR builds now. However for the Nix user, we may want to enable the optimizing profile.

Reply:
Note @zimmi48 that one big downside of the per-package split for developers is that composition is lost. For example, now you can modify a file in lib and coqide will rebuild just that single file and relink. I'm not sure how this is handled by Nix.

Reply:
@ejgallego Nix doesn't support fine-grained compositionality but only at the package level. So splitting packages (if we can ensure the sources are split as well) is actually in favor of compositionality.

Reply:
So for a developer, if I modify a single file, say `lib/queue.ml`, and I want to build `coqide`, then only that file of the Coq package is rebuilt [as happens now with the regular Dune setup]?

Reply:
No the full package, but note that a developer using Nix wouldn't build with `nix-build`. They would use `nix-shell` to get all the dependencies and then would build locally using Dune like any other developer. The compositionality of having several Nix package would only get useful in CI, where `nix-build` is used. A fully nixified CI could support not rebuilding any of the external projects if only the documentation of Coq changed.